feat(config): load user-level context files

Amolith opened the pull request:
Loads `~/.config/AGENTS.md` and `~/.config/crush/CRUSH.md` by default and differentiates them from project context by modifying the coder agent's prompt to refer to them as user preferences. I like this, personally, because project rules might conflict with the user's preferences, and in those cases, the project's rules should take precedence. It also includes the usual docs and config schema bits.

Fixes: https://github.com/charmbracelet/crush/issues/1050
Fixes: https://github.com/charmbracelet/crush/issues/1643
Supersedes: https://github.com/charmbracelet/crush/pull/1668

- [x] I have read [`CONTRIBUTING.md`](https://github.com/charmbracelet/.github/blob/main/CONTRIBUTING.md).
- [x] I have created a discussion that was approved by a maintainer (for new features).

LarsArtmann reviewed (approved):
Looks good 👍

LarsArtmann commented on README.md:
```diff
@@ -334,6 +334,31 @@ using `$(echo $VAR)` syntax.
+
+Crush automatically includes two files for cross-project instructions.
+
+- `~/.config/crush/CRUSH.md`: Crush-specific rules that would confuse other
```
> ~/.config/crush/CRUSH.md

Why not ~/.config/crush/AGENTS.md?

Amolith replied:
Because Crush already uses `~/.config/crush/CRUSH.md` over `~/.config/crush/AGENTS.md`, and if that behaviour is worth changing, I don't think this PR is the place to do it. Those particular lines are just documenting the current behaviour.

meowgorithm commented:
This is a good feature and I find myself wanting it more and more to make Crush behave in ways that are specific to my very personal workflow. That is to say, I think it's time we get it in.

@andreynering wdyt?

andreynering commented on internal/config/load.go:
```diff
@@ -395,6 +395,20 @@ func (c *Config) setDefaults(workingDir, dataDir string) {
 	}
+	if c.Options.ContextPaths == nil {
+		c.Options.ContextPaths = []string{}
+	}
```
This is not needed. `nil` slices behaves the same as an empty slice.


```suggestion
```

andreynering commented on internal/config/load.go:
```diff
@@ -395,6 +395,20 @@ func (c *Config) setDefaults(workingDir, dataDir string) {
+	c.Options.ContextPaths = append(c.Options.ContextPaths, c.Options.MemoryPaths...)
+	if c.Options.SkillsPaths == nil {
+		c.Options.SkillsPaths = []string{}
+	}
```
Same here.


```suggestion
```

andreynering commented on internal/config/config.go:
```diff
@@ -241,6 +241,7 @@ func (Attribution) JSONSchemaExtend(schema *jsonschema.Schema) {
 
 type Options struct {
 	ContextPaths              []string     `json:"context_paths,omitempty" jsonschema:"description=Paths to files containing context information for the AI,example=.cursorrules,example=CRUSH.md"`
+	MemoryPaths               []string     `json:"memory_paths,omitempty" jsonschema:"description=Paths to files containing memory information for the AI,default=~/.config/crush/CRUSH.md,default=~/.config/AGENTS.md"`
```
Why call it "memory"? I think `GlobalContextPaths` would be more clear?

Amolith replied:
I thought it would be a useful distinction to make to the model at the time, but it's probably not worth the less-clear naming. Will change :+1:

andreynering commented on internal/agent/prompt/prompt.go:
```diff
@@ -219,8 +231,11 @@ func (p *Prompt) promptData(ctx context.Context, provider, model string, store *
+	}
+	for _, files := range memoryFiles {
+		data.MemoryFiles = append(data.MemoryFiles, files...)
 	}
```
We should skip adding those for tests. We don't the user `CRUSH.md` to affect the prompt on integration tests. Those are sensitive to changes. If a single character change, tests will fail.

```suggestion
	if !testing.Testing() {
		for _, files := range memoryFiles {
			data.MemoryFiles = append(data.MemoryFiles, files...)
		}
	}
```

andreynering commented on internal/config/load.go:
```diff
@@ -395,6 +395,20 @@ func (c *Config) setDefaults(workingDir, dataDir string) {
+	if c.Options.ContextPaths == nil {
+		c.Options.ContextPaths = []string{}
+	}
+	if c.Options.MemoryPaths == nil {
```
Check for `nil` _or_ empty. It's less error prone.

```suggestion
	if len(c.Options.MemoryPaths) == 0 {
```

andreynering commented on internal/config/load.go:
```diff
@@ -395,6 +395,20 @@ func (c *Config) setDefaults(workingDir, dataDir string) {
+	}
+	if c.Options.MemoryPaths == nil {
+		crushConfigDir := filepath.Dir(GlobalConfig())
+		c.Options.MemoryPaths = []string{
+			filepath.Join(crushConfigDir, "CRUSH.md"),
+			filepath.Join(filepath.Dir(crushConfigDir), "AGENTS.md"),
```
* We should probably support _both_ `CRUSH.md` and `AGENTS.md` here.
* Loading from `~/.config` is strange IMO. _Maybe_ we could load from `$HOME`, but I think `~/.config/crush` is probably enough.

```suggestion
			filepath.Join(crushConfigDir, "CRUSH.md"),
			filepath.Join(crushConfigDir, "AGENTS.md"),
```

Amolith replied:
> Loading from ~/.config is strange IMO. Maybe we could load from $HOME, but I think ~/.config/crush is probably enough.

I agree that it seems strange, but it is a common file for generic, harness-agnostic instructions that other agents look for and Crush would ignore, without either this or user config. Off the top of my head, [Octo](https://github.com/synthetic-lab/octofriend), [Amp](https://ampcode.com/), and [Shelley](https://github.com/boldsoftware/shelley) all source it along with whatever their harness-specific context path is.

I'd _like_ if some form of this remained, but it's configurable so not at all a huge deal at the end of the day.

> We should probably support both CRUSH.md and AGENTS.md here.

~~Will do~~ Have done :+1:

andreynering reviewed (commented):
Hey @Amolith,

Yes, this looks useful. A added a few comments for discussion below.

andreynering commented:
The failures on CI are due to the VCR cassettes. I can re-record those for you, once we get the PR right.

indrora commented:
If I can toss my hat into the ring: A follow-up on this would be to make it possible to specify a set of context files globally -- e.g. something like this:

```json
{
  "options": {
    "instructions": ["~/.crush/CRUSH.md", "~/.config/LLMs.md"]
  }
}
```

Amolith commented:
@indrora would this setting fill that need?  
https://github.com/Amolith/crush/blob/20f191f8d5cd4ec1a340449442915d1ec977e333/README.md#global-context-files

Amolith commented:
@andreynering I think this is ready for another review when convenient! I left https://github.com/charmbracelet/crush/pull/1669#discussion_r3067607118 unresolved because you might have further thoughts there.

meowgorithm commented:
Quick question, @Amolith: these are intended to supplement the project's AGENTS.md (and supersede, where applicable), correct? If so, I agree with the direction and would love to see this one merged.

Side note: let's make sure it supports Windows as well.

Amolith commented:
@meowgorithm yes but I've not given explicit direction around which should supersede another, just effectively said "follow this across all projects".
https://github.com/charmbracelet/crush/pull/1669/changes#diff-267e9e0897afb4256d4160aceb0ba26f78b44257448cae656f33126f9f1de311R412

It _should_ have no trouble on Windows, but I can maybe fire up a VM and test it there.

meowgorithm commented:
Eh, If you're using the same general path resolution as, say, the config you should be good. No need to spin up a VM.

Amolith committed:
feat(config): load user-level context files

Load ~/.config/crush/CRUSH.md and ~/.config/AGENTS.md by default as
global context files, separate from project context. Global context is
presented to the model as user preferences that apply across all
projects. Paths are configurable via global_context_paths.

Amolith committed:
test: update VCR cassettes

meowgorithm commented on schema.json:
```diff
@@ -405,6 +405,17 @@
+          "type": "array",
+          "description": "Paths to files containing global context information for the AI",
+          "default": [
+            "~/.config/crush/CRUSH.md",
+            "~/.config/AGENTS.md"
```
Just double checking: are these valid defaults considering 1) Windows and 2) the builtin Bash interpreter (not sure if its parsing these strings though)?

Amolith replied:
Yes, [`expandPath`](https://github.com/charmbracelet/crush/blob/3e6f95eb2409f5187053d337e6e48dc810b04539/internal/agent/prompt/prompt.go#L138) would turn those into `C:\Users\<user>\.config\AGENTS.md` and `C:\Users\<user>\.config\crush/CRUSH.md`.

But those aren't the actual defaults, just a representation of them. It [uses GlobalConfig()](https://github.com/Amolith/crush/blob/5a180feed6cf61ac9b50b32e7593101b43a97f40/internal/config/load.go#L421-L425) at runtime to get them without translation.
